flash/core/integrations/labelstudio/data_source.py

```diff
@@ -0,0 +1,254 @@
+import json
+import os
+from copy import deepcopy
+from typing import Any, Mapping, Optional, Sequence, TypeVar, Union
+
+import torch
+from pytorch_lightning.trainer.states import RunningStage
+from pytorch_lightning.utilities.cloud_io import get_filesystem
+
+from flash import DataSource
+from flash.core.data.auto_dataset import AutoDataset, IterableAutoDataset
+from flash.core.data.data_source import DefaultDataKeys, has_len
+from flash.core.utilities.imports import _PYTORCHVIDEO_AVAILABLE, _TEXT_AVAILABLE
+
+if _PYTORCHVIDEO_AVAILABLE:
+    from torchvision.datasets.folder import default_loader
+DATA_TYPE = TypeVar("DATA_TYPE")
+
+
+class LabelStudioDataSource(DataSource):
+    """The ``LabelStudioDatasource`` expects the input to
+    :meth:`~flash.core.data.data_source.DataSource.load_data` to be a json export from label studio."""
+
+    def __init__(self):
+        super().__init__()
+        self.results = []
+        self.test_results = []
+        self.val_results = []
+        self.classes = set()
+        self.data_types = set()
+        self.num_classes = 0
+        self._data_folder = ""
+        self._raw_data = {}
+        self.multi_label = False
+        self.split = None
+
+    def load_data(self, data: Optional[Any] = None, dataset: Optional[Any] = None) -> Sequence[Mapping[str, Any]]:
+        """Iterate through all tasks in exported data and construct train\test\val results."""
+        if data and isinstance(data, dict):
+            data_folder = data.get("data_folder")
+            file_path = data.get("export_json")
+            fs = get_filesystem(file_path)
+            with fs.open(file_path) as f:
+                _raw_data = json.load(f)
+            self.multi_label = data.get("multi_label", False)
+            self.split = data.get("split")
+            results, test_results, classes, data_types = LabelStudioDataSource._load_json_data(
+                _raw_data, data_folder=data_folder, multi_label=self.multi_label
+            )
+            self.classes = self.classes | classes
+            self.data_types = self.data_types | data_types
+            self.num_classes = len(self.classes)
+            # splitting result to train and val sets
+            if self.split:
+                import random
+
+                random.shuffle(results)
+                prop = int(len(results) * self.split)
+                self.val_results = results[:prop]
+                self.results = results[prop:]
+                self.test_results = test_results
+                return self.results
+            return results + test_results
+        return []
+
+    def load_sample(self, sample: Mapping[str, Any] = None, dataset: Optional[Any] = None) -> Any:
+        """Load 1 sample from dataset."""
+        # all other data types
+        input_data = deepcopy(sample)
+        try:
```

Comment:
Mind explaining what is done there and add a comment for the future readers.

Reply:
Added a couple of comments